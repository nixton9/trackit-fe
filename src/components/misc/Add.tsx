import React, { useState, useEffect } from 'react'
import Drawer from './Drawer'
import AddNote from '../notes/AddNote'
import AddTask from '../tasks/AddTask'
import AddHabit from '../habits/AddHabit'
import AddExpense from '../expenses/AddExpense'
import { useToggleElement } from '../../utils/useToggleElement'
import { ModuleTypes } from '../../utils/ModuleTypes'
import { ReactComponent as PlusIcon } from '../../assets/icons/plus.svg'
import { ReactComponent as NotesIcon } from '../../assets/icons/notes.svg'
import { ReactComponent as TasksIcon } from '../../assets/icons/tasks.svg'
import { ReactComponent as ExpensesIcon } from '../../assets/icons/expenses.svg'
import { ReactComponent as HabitsIcon } from '../../assets/icons/habits.svg'
import { ReactComponent as ChevronIcon } from '../../assets/icons/chevron.svg'
import { Styled } from '../../styles/Add.styles'

const notesTitle = 'Create a note'
const tasksTitle = 'Create a task'
const expensesTitle = 'Add an expense'
const habitsTitle = 'Add an habit'

export const AddSubmitButton: React.FC<{ handleSubmit: () => void }> = ({
  handleSubmit
}) => (
  <Styled.AddWidget__Button onClick={handleSubmit}>
    <ChevronIcon />
  </Styled.AddWidget__Button>
)

const Add: React.FC = () => {
  const [activeContent, setActiveContent] = useState<ModuleTypes | null>(null)
  const [open, setOpen, overlayEl] = useToggleElement()

  useEffect(() => {
    if (open) setActiveContent(null)
  }, [open])

  let drawerTitle
  let drawerContent

  switch (activeContent) {
    case ModuleTypes.Notes:
      drawerTitle = notesTitle
      drawerContent = <AddNote closeModal={() => setOpen(false)} />
      break

    case ModuleTypes.Tasks:
      drawerTitle = tasksTitle
      drawerContent = <AddTask />
      break

    case ModuleTypes.Expenses:
      drawerTitle = expensesTitle
      drawerContent = <AddExpense />
      break

    case ModuleTypes.Habits:
      drawerTitle = habitsTitle
      drawerContent = <AddHabit />
      break

    default:
      drawerTitle = 'What do you want to track?'
      drawerContent = (
        <>
          <Styled.AddButton onClick={() => setActiveContent(ModuleTypes.Notes)}>
            <span>{notesTitle}</span>
            <NotesIcon />
          </Styled.AddButton>

          <Styled.AddButton onClick={() => setActiveContent(ModuleTypes.Tasks)}>
            <span>{tasksTitle}</span>
            <TasksIcon />
          </Styled.AddButton>

          <Styled.AddButton
            onClick={() => setActiveContent(ModuleTypes.Expenses)}
          >
            <span>{expensesTitle}</span>
            <ExpensesIcon />
          </Styled.AddButton>

          <Styled.AddButton
            onClick={() => setActiveContent(ModuleTypes.Habits)}
          >
            <span>{habitsTitle}</span>
            <HabitsIcon />
          </Styled.AddButton>
        </>
      )
  }

  return (
    <>
      <Styled.AddIcon onClick={() => setOpen(true)}>
        <PlusIcon />
      </Styled.AddIcon>

      <Drawer title={drawerTitle} open={open} overlayRef={overlayEl}>
        {drawerContent}
      </Drawer>
    </>
  )
}

export default Add
