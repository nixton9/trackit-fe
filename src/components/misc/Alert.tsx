import React from 'react'
import { Styled } from '../../styles/Notifications.styles'
import { useRecoilState, atom } from 'recoil'

export const alertState: any = atom({
  key: 'alert',
  default: {
    text: '',
    onConfirm: undefined
  }
})

export const Alert: React.FC = () => {
  const [alert, setAlert] = useRecoilState(alertState)

  const { text, onConfirm } = alert

  if (text && onConfirm) {
    const close = () => {
      setAlert({ text: '', onConfirm: undefined })
    }

    const handleConfirm = () => {
      onConfirm()
      close()
    }

    return (
      <Styled.NotificationContainer>
        <Styled.Alert>
          <p>{text}</p>
          <Styled.Alert__Buttons>
            <button className="confirm" onClick={handleConfirm}>
              Do it
            </button>
            <button onClick={close}>Cancel</button>
          </Styled.Alert__Buttons>
        </Styled.Alert>
      </Styled.NotificationContainer>
    )
  } else {
    return null
  }
}
