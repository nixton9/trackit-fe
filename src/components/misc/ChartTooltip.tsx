import React from 'react'
import { Styled } from '../../styles/Charts.styles'

type ChartTooltipProps = {
  active: boolean
  payload: any
  label: string
}

export const ChartTooltip: React.FC<ChartTooltipProps> = ({
  active,
  payload,
  label
}) => {
  if (active) {
    const dataKey = payload[0].dataKey

    return (
      <Styled.Tooltip>
        <h5>
          {payload[0].payload.monthLabel || payload[0].payload.name || label}
        </h5>
        <h4>
          {payload[0].payload.displayValue || payload[0].payload[dataKey]}
        </h4>
        <p>Total spent</p>
      </Styled.Tooltip>
    )
  }
  return null
}
