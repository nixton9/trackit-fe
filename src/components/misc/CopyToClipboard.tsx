import React, { useRef } from 'react'
import styled from 'styled-components'

const TextArea = styled.textarea`
  position: absolute;
  top: -9999px;
  left: -9999px;
  opacity: 0;
`

type CopyToClipboardProps = {
  text: string
  onCopy: () => void
}

export const CopyToClipboard: React.FC<CopyToClipboardProps> = ({
  text,
  onCopy
}) => {
  const textAreaRef = useRef<HTMLTextAreaElement>(null)
  const copyToClipboard = (e: React.MouseEvent<HTMLDivElement>) => {
    textAreaRef.current?.select()
    document.execCommand('copy')
    onCopy()
  }

  return (
    <>
      <div onClick={copyToClipboard}>Copy text</div>
      <TextArea ref={textAreaRef} value={text} onChange={() => null} />
    </>
  )
}
