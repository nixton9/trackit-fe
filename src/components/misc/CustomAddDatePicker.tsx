import React, { forwardRef } from 'react'
import { ReactComponent as CalendarIcon } from '../../assets/icons/calendr.svg'

export const CustomAddDatePicker = forwardRef(
  ({ onClick, value, onChange }: any, ref: any) => (
    <>
      <CalendarIcon onClick={onClick} />
      <input
        type="text"
        value={value}
        onClick={onClick}
        onChange={onChange}
        ref={ref}
      />
    </>
  )
)
