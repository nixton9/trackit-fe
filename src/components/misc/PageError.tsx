import React from 'react'
import { Styled } from '../../styles/Page.styles'
import { ReactComponent as ErrorIcon } from '../../assets/icons/error.svg'

export const PageError: React.FC = ({ children }) => (
  <Styled.PageError>
    <ErrorIcon />
    {children}
  </Styled.PageError>
)
