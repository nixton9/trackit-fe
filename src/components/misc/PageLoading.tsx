import React from 'react'
import { LoadingSpinner } from './LoadingSpinner'
import { Styled } from '../../styles/Page.styles'

type PageLoadingProps = {
  centered?: boolean
}

export const PageLoading: React.FC<PageLoadingProps> = ({ centered }) => (
  <Styled.PageLoading className={centered ? 'centered' : ''}>
    <LoadingSpinner />
  </Styled.PageLoading>
)
