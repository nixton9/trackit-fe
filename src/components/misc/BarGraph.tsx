import React from 'react'
import { ChartTooltip } from './ChartTooltip'
import { Styled } from '../../styles/Charts.styles'
import { BarChart, Bar, XAxis, Tooltip, ResponsiveContainer } from 'recharts'

type BarGraphProps = {
  data: {
    name: string
    value: number | string
    monthLabel?: string
    displayValue?: string
  }[]
  bars: { key: string; color: string }[]
}

export const BarGraph: React.FC<BarGraphProps> = ({ data, bars }) => {
  const chartData = data.map(item => ({
    ...item,
    value: item.value === 0 ? 1 : item.value
  }))

  return (
    <Styled.GraphContainer className="bar">
      <ResponsiveContainer>
        <BarChart width={500} height={300} data={chartData}>
          <XAxis dataKey="name" />
          <Tooltip content={ChartTooltip} />
          {bars &&
            bars.length &&
            bars.map(bar => (
              <Bar
                key={bar.key}
                dataKey={bar.key}
                fill={bar.color}
                radius={4}
              />
            ))}
        </BarChart>
      </ResponsiveContainer>
    </Styled.GraphContainer>
  )
}
