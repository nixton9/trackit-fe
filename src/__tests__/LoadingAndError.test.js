import React from 'react'
import { render } from '@testing-library/react'
import { PageLoading } from '../components/misc/PageLoading'
import { PageError } from '../components/misc/PageError'

describe('Loading', () => {
  it('matches snapshot', () => {
    const { container } = render(<PageLoading />)

    expect(container).toMatchSnapshot()
  })
})

describe('Error', () => {
  it('matches snapshot', () => {
    const { container } = render(<PageError>Test</PageError>)

    expect(container).toMatchSnapshot()
  })

  it('shows the error message', () => {
    const { getByText } = render(<PageError>Error fetching data</PageError>)

    expect(getByText('Error fetching data')).toBeInTheDocument()
  })
})
