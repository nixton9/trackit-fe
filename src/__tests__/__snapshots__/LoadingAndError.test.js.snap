// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error matches snapshot 1`] = `
<div>
  <div
    class="Pagestyles__PageError-sc-19pj6gm-14 hfccEd"
  >
    <svg>
      error.svg
    </svg>
    <p>
      Test
    </p>
  </div>
</div>
`;

exports[`Loading matches snapshot 1`] = `
<div>
  <div
    class="Pagestyles__PageLoading-sc-19pj6gm-13 feDNhr"
  >
    <div
      class="sc-AxjAm gcvfvo loading-spinner"
    >
      <div
        class="sc-AxirZ emeNQI"
      />
      <div
        class="sc-AxirZ emeNQI doublebounce2"
      />
    </div>
  </div>
</div>
`;
